storage/aptosdb/src/fast_sync_aptos_db.rs

```diff
@@ -0,0 +1,1110 @@
+use std::sync::{Arc, RwLock};
+
+const SECONDARY_DB_DIR: &str = "fast_sync_secondary";
+
+/// This is a wrapper around [AptosDB] that is used to bootstrap the node for fast sync mode
+pub struct FastSyncStorageWrapper {
+    // main is used for normal read/write or genesis data during fast sync
+    db_main: AptosDB,
+    // secondary is used for restoring fast sync snapshot and all the read/writes afterwards
+    db_secondary: Option<AptosDB>,
+    // This is for reading the fast_sync status to determine which db to use
+    fast_sync_status: Arc<RwLock<FastSyncStatus>>,
+}
+
+impl FastSyncStorageWrapper {
+    pub fn new_fast_sync_aptos_db(
+        config: &NodeConfig,
+        status: Arc<RwLock<FastSyncStatus>>,
+    ) -> Result<Self> {
+        let mut db_dir = config.storage.dir();
+        let db_main = AptosDB::open(
+            db_dir.as_path(),
+            false,
+            config.storage.storage_pruner_config,
+            config.storage.rocksdb_configs,
+            config.storage.enable_indexer,
+            config.storage.buffered_state_target_items,
+            config.storage.max_num_nodes_per_lru_cache_shard,
+        )
+        .map_err(|err| anyhow!("fast sync DB failed to open {}", err))?;
+
+        let db_secondary = if config.state_sync.state_sync_driver.bootstrapping_mode
+            == BootstrappingMode::DownloadLatestStates
+        {
+            db_dir.push(SECONDARY_DB_DIR);
+            Some(
+                AptosDB::open(
+                    db_dir.as_path(),
+                    false,
+                    config.storage.storage_pruner_config,
+                    config.storage.rocksdb_configs,
+                    config.storage.enable_indexer,
+                    config.storage.buffered_state_target_items,
+                    config.storage.max_num_nodes_per_lru_cache_shard,
+                )
+                .map_err(|err| anyhow!("fast sync DB failed to open {}", err))?,
+            )
+        } else {
+            None
+        };
+
+        Ok(Self {
+            db_main,
+            db_secondary,
+            fast_sync_status: status,
+        })
+    }
+
+    pub fn get_fast_sync_status(&self) -> Result<FastSyncStatus> {
+        self.fast_sync_status
+            .read()
+            .map_err(|err| anyhow!("failed to read fast sync status: {}", err))
+            .map(|status| *status)
+    }
+
+    /// Check if the fast sync finished already
+    fn is_fast_sync_bootstrap_finished(&self) -> bool {
+        let status = self.get_fast_sync_status().unwrap();
+        status == FastSyncStatus::FINISHED
+    }
+
+    pub(crate) fn get_transaction_store(&self) -> Result<&TransactionStore> {
+        if self.is_fast_sync_bootstrap_finished() {
+            Ok(&self
+                .db_secondary
+                .as_ref()
+                .expect("db_secondary is not initialized")
+                .transaction_store)
+        } else {
+            Ok(&self.db_main.transaction_store)
+        }
```

Comment:
Too much duplication. We can at least abstract this piece out as `fn get_aptosdb_db_ref() -> &AptosDB`.  I would go as far as doing a 
```
trait GetAptosDbRef {
    fn get_aptos_db_ref() -> &AptosDB;
}
```
And implement that for both the actual AptosDB and the wrapper, and then
```
impl<T: GetAptosDbRef> DbWriter for T {
...
}
```

Reply:
Or darker (but elegant) magic like https://github.com/aptos-labs/aptos-core/blob/547ad67a0126f9447ef9aaf19863a56203a3e17d/aptos-move/aptos-gas-profiling/src/profiler.rs#L51